### ISSUE TITLE
pwngdb no break when debug on chip

### ISSUE THREAD

**Issue by AnduinBrian** [pwndbg/pwndbg#1953]
### Description
Hello, after use openocd to create a local gdb server, i try to attach but pwngdb cant break, it just hang. I try again with normal-gdb, it work.

### Steps to reproduce
I use ST-link and stm32, so i start the server with command 
```
openocd -f /usr/share/openocd/scripts/interface/stlink.cfg -f /usr/share/openocd/scripts/target/stm32f1x.cfg
```
I start gdb-multiarch, and type those command:
```
file firmware.elf                           <- this is my firmware i want to debug
b * loop                                    <- place some bp
target extended-remote localhost:3333       <- openocd port
```
GDB just hang there, i cant interact or do anything with the console.

![image](https://github.com/pwndbg/pwndbg/assets/36282595/63197489-d756-43df-85ba-a3b9f430d5e4)

But when i try with normal-gdb, i can work with it

![image](https://github.com/pwndbg/pwndbg/assets/36282595/a85f188b-c24e-402d-af24-32657cad22b4)

**Comment by disconnect3d** [comment]
Eh, I don't have such hardware to test this myself. We invoke some GDB hooks when e.g. a library is loaded, on program start or when a stop occurs.

Most likely, it is related to us trying to figure out memory mappings and our aggresive reading of parts of memory to provide great experience.

This can be debugged to some degree with `set debug-events on`. I'd be happy to merge improvements that would make chip/embedded debugging more pleasant.

**Comment by nikvoid** [comment]
This is output with `debug-events` and `verbose-exceptions` enabled. Pwndbg just freezes waiting for some events?

```
pwndbg> ctx
// Just stuck until ctrl+C
^CQuit
'before_prompt' pwndbg.lib.cache.clear ()
'stop' pwndbg.lib.cache.clear ()
'stop' pwndbg.dbg.gdb._bpwp_process_deferred ()
'stop' pwndbg.aglib.regs.update_last ()
'stop' pwndbg.aglib.stack.is_executable ()
// Stuck again
^CTraceback (most recent call last):
  File "/home/nv/pwndbg/pwndbg/gdblib/prompt.py", line 95, in prompt_hook
    pwndbg.gdblib.events.after_reload(start=cur is None)
  File "/home/nv/pwndbg/pwndbg/gdblib/events.py", line 365, in after_reload
    invoke_event(gdb.events.stop)
  File "/home/nv/pwndbg/pwndbg/gdblib/events.py", line 360, in invoke_event
    f(*args, **kwargs)
  File "/home/nv/pwndbg/pwndbg/gdblib/events.py", line 287, in caller
    func()
  File "/home/nv/pwndbg/pwndbg/lib/cache.py", line 162, in decorator
    value = func(*a, **kw)
  File "/home/nv/pwndbg/pwndbg/aglib/stack.py", line 72, in is_executable
    ehdr = pwndbg.aglib.elf.exe()
  File "/home/nv/pwndbg/pwndbg/aglib/proc.py", line 151, in wrapper
    return func(*a, **kw)
  File "/home/nv/pwndbg/pwndbg/lib/cache.py", line 162, in decorator
    value = func(*a, **kw)
  File "/home/nv/pwndbg/pwndbg/aglib/elf.py", line 246, in exe
    e = entry()
  File "/home/nv/pwndbg/pwndbg/aglib/proc.py", line 151, in wrapper
    return func(*a, **kw)
  File "/home/nv/pwndbg/pwndbg/lib/cache.py", line 162, in decorator
    value = func(*a, **kw)
  File "/home/nv/pwndbg/pwndbg/aglib/elf.py", line 259, in entry
    entry = pwndbg.auxv.get().AT_ENTRY
  File "/home/nv/pwndbg/pwndbg/lib/cache.py", line 162, in decorator
    value = func(*a, **kw)
  File "/home/nv/pwndbg/pwndbg/auxv.py", line 123, in get
    return use_info_auxv() or walk_stack() or AUXV()
  File "/home/nv/pwndbg/pwndbg/auxv.py", line 176, in walk_stack
    auxv = walk_stack2(0)
  File "/home/nv/pwndbg/pwndbg/auxv.py", line 210, in walk_stack2
    end = find_stack_boundary(sp)
  File "/home/nv/pwndbg/pwndbg/auxv.py", line 162, in find_stack_boundary
    if b"\x7fELF" == pwndbg.aglib.memory.read(addr, 4):
  File "/home/nv/pwndbg/pwndbg/aglib/memory.py", line 35, in read
    return pwndbg.dbg.selected_inferior().read_memory(address=addr, size=count, partial=partial)
  File "/home/nv/pwndbg/pwndbg/dbg/gdb/__init__.py", line 407, in read_memory
    result = gdb.selected_inferior().read_memory(addr, count)
KeyboardInterrupt: Quit
pwndbg>
```
EDIT: Further investigation with debug print inserted at [pwndbg/pwndbg/aglib/memory.py](https://github.com/pwndbg/pwndbg/blob/9f1753f4d0558fd4d229750a1d8c85ee7c8b704b/pwndbg/aglib/memory.py#L35) shows that it tries to read words of memory from a ton of addresses

```
...
20040000 @ 4, partial: False
20041000 @ 4, partial: False
20042000 @ 4, partial: False
20043000 @ 4, partial: False
20044000 @ 4, partial: False
20045000 @ 4, partial: False
20046000 @ 4, partial: False
20047000 @ 4, partial: False
20048000 @ 4, partial: False
20049000 @ 4, partial: False
2004A000 @ 4, partial: False
2004B000 @ 4, partial: False
2004C000 @ 4, partial: False
2004D000 @ 4, partial: False
2004E000 @ 4, partial: False
2004F000 @ 4, partial: False
20050000 @ 4, partial: False
20051000 @ 4, partial: False
20052000 @ 4, partial: False
20053000 @ 4, partial: False
20054000 @ 4, partial: False
20055000 @ 4, partial: False
20056000 @ 4, partial: False
20057000 @ 4, partial: False
...
```
So, the questions is: how to disable such aggressive scan?

**Comment by patryk4815** [comment]
remove. walk_stack

**Comment by patryk4815** [comment]
I had same problem, but with Golang binary , on linux
<img width="525" alt="Screenshot 2024-12-18 at 21 49 56" src="https://github.com/user-attachments/assets/e249761b-5fc9-4b4c-9687-678aaa93236b" />

**Comment by patryk4815** [comment]
@nikvoid Please check If PR https://github.com/pwndbg/pwndbg/pull/2641 resolve your issue

**Comment by nikvoid** [comment]
It seems to be better - some other command like `u $pc` are working now, but `context` still stucks in `read_memory`.

```
pwndbg> ctx
^CQuit
'before_prompt' pwndbg.lib.cache.clear ()
^CTraceback (most recent call last):
  File "/home/nv/pwndbg/pwndbg/gdblib/prompt.py", line 100, in prompt_hook
    pwndbg.commands.context.context()
  File "/home/nv/pwndbg/pwndbg/commands/__init__.py", line 183, in __call__
    return self.function(*args, **kwargs)
  File "/home/nv/pwndbg/pwndbg/commands/context.py", line 658, in context
    func(
  File "/home/nv/pwndbg/pwndbg/commands/context.py", line 327, in _serve_context_history
    current_output = function(*a, **kw)
  File "/home/nv/pwndbg/pwndbg/commands/context.py", line 767, in context_regs
    regs = get_regs()
  File "/home/nv/pwndbg/pwndbg/commands/context.py", line 862, in get_regs
    desc = pwndbg.chain.format(value)
  File "/home/nv/pwndbg/pwndbg/chain.py", line 141, in format
    rest = [M.get_address_and_symbol(link) for link in chain]
  File "/home/nv/pwndbg/pwndbg/chain.py", line 141, in <listcomp>
    rest = [M.get_address_and_symbol(link) for link in chain]
  File "/home/nv/pwndbg/pwndbg/color/memory.py", line 36, in get_address_and_symbol
    page = pwndbg.aglib.vmmap.find(address)
  File "/home/nv/pwndbg/pwndbg/lib/cache.py", line 162, in decorator
    value = func(*a, **kw)
  File "/home/nv/pwndbg/pwndbg/aglib/vmmap.py", line 38, in find
    return pwndbg.gdblib.vmmap.explore(address)
  File "/home/nv/pwndbg/pwndbg/gdblib/abi.py", line 65, in caller
    return func(*args, **kwargs)
  File "/home/nv/pwndbg/pwndbg/gdblib/vmmap.py", line 266, in explore
    page = find_boundaries(address_maybe)
  File "/home/nv/pwndbg/pwndbg/gdblib/vmmap.py", line 737, in find_boundaries
    start = pwndbg.aglib.memory.find_lower_boundary(addr)
  File "/home/nv/pwndbg/pwndbg/lib/cache.py", line 162, in decorator
    value = func(*a, **kw)
  File "/home/nv/pwndbg/pwndbg/aglib/memory.py", line 338, in find_lower_boundary
    read(addr, 1)
  File "/home/nv/pwndbg/pwndbg/aglib/memory.py", line 35, in read
    return pwndbg.dbg.selected_inferior().read_memory(address=addr, size=count, partial=partial)
  File "/home/nv/pwndbg/pwndbg/dbg/gdb/__init__.py", line 407, in read_memory
    result = gdb.selected_inferior().read_memory(addr, count)
KeyboardInterrupt: Quit
```


Again, `read_memory` debug-prints:
```
pwndbg> ctx
0 @ 1
0 @ 1
0 @ 1
1000 @ 1
2000 @ 1
3000 @ 1
4000 @ 1
5000 @ 1
6000 @ 1
7000 @ 1
8000 @ 1
9000 @ 1
A000 @ 1
B000 @ 1
C000 @ 1
D000 @ 1
E000 @ 1
F000 @ 1
10000 @ 1
...
// But it works after all
...
20031D48 @ 1
20031D4C @ 1
20031D50 @ 1
20031D54 @ 1
LEGEND: STACK | HEAP | CODE | DATA | WX | RODATA
───────────[ REGISTERS / show-flags off / show-compact-regs off ]───────────
 R0   0 —▸ 0xd7486e0b ◂— 0
 R1   0 —▸ 0xd7486e0b ◂— 0
 R2   0x4001340c —▸ 0xffd9 —▸ 0 —▸ 0xd7486e0b ◂— 0
 R3   1 —▸ 0x79d7486e —▸ 0 —▸ 0xd7486e0b ◂— 0
 R4   5 —▸ 0xd47fe37c —▸ 0 —▸ 0xd7486e0b ◂— 0
 R5   109 —▸ 0x3267db58 —▸ 0 —▸ 0xd7486e0b ◂— 0
 R6   61 —▸ 0xf56935cb —▸ 0 —▸ 0xd7486e0b ◂— 0
 R7   56 —▸ 0xc8e7e587 —▸ 0 —▸ 0xd7486e0b ◂— 0
 R8   0 —▸ 0xd7486e0b ◂— 0
 R9   4 —▸ 0x7fe37c79 —▸ 0 —▸ 0xd7486e0b ◂— 0
 R10  2 —▸ 0x7c79d748 —▸ 0 —▸ 0xd7486e0b ◂— 0
 R11  45 —▸ 0x54f2099a —▸ 0 —▸ 0xd7486e0b ◂— 0
 R12  0xa5a5a5a5 —▸ 0 —▸ 0xd7486e0b ◂— 0
 SP   0x20031d38 —▸ 207 —▸ 0x7488e369 —▸ 0 —▸ 0xd7486e0b ◂— ...
 LR   0x8037661 —▸ 0x600849 —▸ 0 —▸ 0xd7486e0b ◂— 0
 PC   0x802b3e8 —▸ 0x47f0e92d —▸ 0 —▸ 0xd7486e0b ◂— 0
──────────────[ DISASM / armcm / thumb mode / set emulate on ]──────────────
 ► 0x802b3e8    push.w {r4, r5, r6, r7, r8, sb, sl, lr}
   0x802b3ec    mov    r4, r0                               R4 => 0 —▸ 0xd7486e0b ◂— 0
   0x802b3ee    mov    r5, r1                               R5 => 0 —▸ 0xd7486e0b ◂— 0
   0x802b3f0    movs   r7, #0                               R7 => 0 —▸ 0xd7486e0b ◂— 0
   0x802b3f2    mov    r8, r7                               R8 => 0 —▸ 0xd7486e0b ◂— 0
   0x802b3f4    mov    sb, r7                               SB => 0 —▸ 0xd7486e0b ◂— 0
   0x802b3f6    movs   r6, #0                               R6 => 0 —▸ 0xd7486e0b ◂— 0
   0x802b3f8    mov    sl, r6                               SL => 0 —▸ 0xd7486e0b ◂— 0
   0x802b3fa    cmp    r4, #7                               0 - 7     XPSR => 0x81000000 [ N z c v q T ]
   0x802b3fc  ✔ bls    #0x802b404                  <0x802b404>
    ↓
   0x802b404    cbnz   r5, #0x802b418              <0x802b418>
─────────────────────────────────[ STACK ]──────────────────────────────────
00:0000│ sp 0x20031d38 —▸ 207 —▸ 0x7488e369 —▸ 0 —▸ 0xd7486e0b ◂— ...
01:0004│    0x20031d3c —▸ 0 —▸ 0xd7486e0b ◂— 0
02:0008│    0x20031d40 —▸ 10 —▸ 0x29432be5 —▸ 0 —▸ 0xd7486e0b ◂— ...
03:000c│    0x20031d44 —▸ 61 —▸ 0xf56935cb —▸ 0 —▸ 0xd7486e0b ◂— ...
04:0010│    0x20031d48 —▸ 111 —▸ 0xe7723267 —▸ 0 —▸ 0xd7486e0b ◂— ...
05:0014│    0x20031d4c —▸ 172 —▸ 0xe308d50 —▸ 0 —▸ 0xd7486e0b ◂— ...
06:0018│    0x20031d50 —▸ 0x2c8d —▸ 0x2a857212 —▸ 0 —▸ 0xd7486e0b ◂— ...
07:001c│    0x20031d54 —▸ 0x190f —▸ 0x65bfef68 —▸ 0 —▸ 0xd7486e0b ◂— ...
───────────────────────────────[ BACKTRACE ]────────────────────────────────
 ► 0 0x802b3e8 None
   1 0x8037660 None
────────────────────────────────────────────────────────────────────────────
```

EDIT: Looking at the call stack, problem with aggressive scan persists. Vmmap for embedded devices could be retrieved with `gdb info mem`:
```
pwndbg> info mem
Using memory regions provided by the target.
Num Enb Low Addr   High Addr  Attrs
0   y   0x00000000 0x00200000 rw nocache
1   y   0x00200000 0x00220000 flash blocksize 0x8000 nocache
2   y   0x00220000 0x00240000 flash blocksize 0x20000 nocache
3   y   0x00240000 0x00300000 flash blocksize 0x40000 nocache
4   y   0x00300000 0x08000000 rw nocache
5   y   0x08000000 0x08020000 flash blocksize 0x8000 nocache
6   y   0x08020000 0x08040000 flash blocksize 0x20000 nocache
7   y   0x08040000 0x08100000 flash blocksize 0x40000 nocache
8   y   0x08100000 0x1ff0f000 rw nocache
9   y   0x1ff0f000 0x1ff0f400 flash blocksize 0x40 nocache
10  y   0x1ff0f400 0x100000000 rw nocache
```

**Comment by disconnect3d** [comment]
Huh, TIL `info mem`. I guess we could fallback to using it in `vmmap` when we don't have memory mappings.

The problem is that we explore vmmap mappings at all by default. We probably should never do this by default and instead tell the user "hey, no vmmap info is available: you either set vmmaps yourself, OR, use this magic command or setting to do the exploration" (either explore once and cache forever, or, re-explore automagically? I am not sure).

TL;DR simplified stack trace:

1. Context is displayed
2. It want to enhance the display of a value, e.g., in a register via a call to the `pwndbg.chain.format(value)`
3. This goes to `pwndbg/chain.py:141    M.get_address_and_symbol(link)`
4. `pwndbg/color/memory.py:36    page = pwndbg.aglib.vmmap.find(address)`
5. caching layer
6. `pwndbg/aglib/vmmap.py:38    return pwndbg.gdblib.vmmap.explore(address)`
7. `find_boundaries` -> `pwndbg.aglib.memory.find_lower_boundary -> read(addr, 1)`

**Comment by disconnect3d** [comment]
Fwiw we have a bunch of vmmap related commands already:
```
pwndbg> pwndbg vmmap
Memory Commands [Aliases]         Description
--------------------------------  --------------------------------------------
vmmap [lm, address, vprot, libs]  Print virtual memory map pages.
vmmap_add                         Add virtual memory map page.
vmmap_clear                       Clear the vmmap cache.
vmmap_explore                     Explore a page, trying to guess permissions.
vmmap_load                        Load virtual memory map pages from ELF file.
```

You should be able to do `vmmap_add ...` or `vmmap_load` to add custom vmmaps and then exploration should not occur I think.

**Comment by disconnect3d** [comment]
@nikvoid Regarding `info mem` - is there any way we can determine if a page is executable? Are all pages executable? Can we determine any names for them (like which one is code/binary, stack/heap, etc)?

Do you know if `info mem` have consistent output format, or is it always as shown in your listing?

**Comment by nikvoid** [comment]
Well, I found this command in the other issue in this repo related to embedded debugging)
So I don't know exactly. In this particular output, `flash`  is RO memory, and `rw` is microcontroller RAM and also peripheral registers. The problem though, RAM address space is indistinguishable from vast address space of registers. I think this is fine if you don't poke at it aggressively and just use address bounds to highlight R/W/X data.
Regarding stack/heap - it is determined by the linker scripts, RAM is combined from static variables + stack (maybe multiple stacks if you are using some RTOS) + heap, which may be either some large static array or named section in linker script.

**Comment by nikvoid** [comment]
BTW I tried using `vmmap_load` on firmware ELF and it does not prevent scanning

**Comment by nikvoid** [comment]
I would like to reopen as issue persists

**Comment by patryk4815** [comment]
@nikvoid This issue has been fixed in https://github.com/pwndbg/pwndbg/pull/2646 . 
Please verify the fix, and feel free to reopen the issue if you encounter any further problems.

**Comment by nikvoid** [comment]
Yeah, now it's way usable. Thank you for contributions!

Is there a way to specify memory segment type? E.g. I can use `vmmap_add` to add stack range, but how to make it recognized as "stack"?

**Comment by patryk4815** [comment]
@nikvoid use command `stack_explore` + `vmmap_load ./your/file.elf`. I think there is no way to specify stack or not in `vmmap_add`. If you see something to improve please create another feature ticket :)
